lib/SPIRV/SPIRVOps.cpp

```diff
@@ -740,6 +740,102 @@ static LogicalResult verify(spirv::VariableOp varOp) {
+    int32_t index = indexAttr.dyn_cast<IntegerAttr>().getInt();
+    if (resultType.getKind() == spirv::TypeKind::Array) {
+      if (index < 0 ||
+          static_cast<uint32_t>(index) >=
+              resultType.cast<spirv::ArrayType>().getElementCount()) {
+        parser->emitError(attrLocation, "index out of bounds");
+        return nullptr;
+      }
+      resultType = resultType.cast<spirv::ArrayType>().getElementType();
+    } else if (resultType.getKind() == spirv::TypeKind::Struct) {
+      if (index < 0 ||
+          static_cast<uint32_t>(index) >=
+              resultType.cast<spirv::StructType>().getNumMembers()) {
+        parser->emitError(attrLocation, "index out of bounds");
+        return nullptr;
+      }
+      resultType = resultType.cast<spirv::StructType>().getMemberType(index);
+    } else if (resultType.getKind() == StandardTypes::Vector) {
+      if (index < 0 || static_cast<uint32_t>(index) >=
+                           resultType.cast<VectorType>().getNumElements()) {
+        parser->emitError(attrLocation, "index out of bounds");
+        return nullptr;
+      }
+      resultType = resultType.cast<VectorType>().getElementType();
+    } else {
+      parser->emitError(attrLocation, "invalid type for indexing");
+      return nullptr;
+    }
+  }
+  return resultType;
+}
+
+static ParseResult parseCompositeExtractOp(OpAsmParser *parser,
+                                           OperationState *state) {
+  OpAsmParser::OperandType compositeInfo;
+  Attribute indicesAttr;
+  Type compositeType;
+  llvm::SMLoc attrLocation;
+
+  if (parser->parseOperand(compositeInfo) ||
+      parser->getCurrentLocation(&attrLocation) ||
+      parser->parseAttribute(indicesAttr, "indices", state->attributes) ||
+      parser->parseColon() || parser->parseType(compositeType) ||
+      parser->resolveOperand(compositeInfo, compositeType, state->operands)) {
+    return failure();
+  }
+
+  auto indicesArrayAttr = indicesAttr.dyn_cast<ArrayAttr>();
+  if (!indicesArrayAttr) {
+    parser->emitError(attrLocation, "expected an array attribute");
+    return failure();
+  }
+
+  Type resultType = getElementTypeByIndices(parser, state, compositeType,
+                                            indicesArrayAttr, attrLocation);
+  if (!resultType) {
+    return failure();
+  }
+
+  state->addTypes(resultType);
+  return success();
+}
+
+static void print(spirv::CompositeExtractOp compositeExtractOp,
+                  OpAsmPrinter *printer) {
+  *printer << spirv::CompositeExtractOp::getOperationName() << ' '
+           << *compositeExtractOp.composite() << compositeExtractOp.indices()
+           << " : " << compositeExtractOp.composite()->getType();
+}
+
+static LogicalResult verify(spirv::CompositeExtractOp compositeExtractOp) {
+  auto resultType = compositeExtractOp.getType();
+  if (resultType == compositeExtractOp.composite()->getType()) {
+    return compositeExtractOp.emitOpError(
```

Comment:
Could you still call the function to verify the result type agree with the extracted type? It's a bit unfortunate that we are duplicating the check logic but we need that in the parser to properly set the result type. And this verifier is used across all paths (parsing custom assembly form, parsing generic assembly form, and building ops with C++ API, etc.) so it would be nice to guarantee.

Reply:
Added a verifier.